veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,374 @@
+
+To track previous backups and enable incremental backups, a new `VirtualMachineBackupChain` CR will be introduced. This CR will maintain the history of checkpoints for a given backup chain. Since a VM might be backed up by multiple solutions, a separate `VirtualMachineBackupChain` will be need to be created to track each chain.
+The `VirtualMachineBackupChain` can be referenced in the `VirtualMachineBackup` CR. Doing so allows checkpoint metadata to be stored and used for the next incremental backup. The chain CR will also include the name of the most recent checkpoint to be used as the base for the next incremental backup. This field is automatically updated at the end of each successful backup. If needed, users may override this field to initiate an incremental backup from a different checkpoint.
+
+If a backup is initiated without referencing a `VirtualMachineBackupChain`, a full backup will be performed, and no checkpoint history will be retained for future incremental backups. If a chain is referenced but contains no checkpoint history, the backup will also default to a full backup.
+
+**Handling Checkpoints Across VM Restarts**
+One challenge in Kubevirt is that Libvirt is re-created each time a VM restarts, causing all existing libvirt checkpoints to be lost. To enable incremental backups after a restart—rather than falling back to a full backup, checkpoints metadata must be provided to libvirt during VM initialization. This allows libvirt to redefine the checkpoints, ensuring that an incremental backup can be continued as expected.
+To address this, the VM controller will use the stored metadata in the VirtualMachineBackupChain CRs during startup. It will iterate through all relevant chains and re-create the checkpoints in Libvirt based on the stored metadata. This ensures that incremental backups can resume seamlessly, even after a VM restart.
+
+### Offline backups
+
+An initial proposed solution involves starting the VM in a paused state to perform the backup, and then either unpausing or shutting it down once the backup is complete. However, this approach introduces additional considerations and trade-offs, which justify a separate, dedicated discussion. As such, this section is intentionally left without a formal API definition for now and only online backup will be supported at first.
+
+### State interruptions during backup
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup or before stopping the VM we would wait for the VirtualMachineBackup job to complete and only then allow the VM to stop.
```

Comment:
Is this fix going to be ready before this feature ships? My concern is that, unlike with VM snapshots, a backup may take a while, especially a full one. Is there a way to gracefully cancel the backup and leave the bitmap in a good state for when the VM comes back up?

Reply:
@pipo correct me if Im wrong, there is an option to send an abort action to the backup job, and we can do it in case the backup object is deleted before the backup is `done`.
Regarding the bitmaps Im not sure what will be the state of it, if the backup never completed will it be deleted automatically, or will need to also delete it after aborting the backup, and then shutdown the VM should be good.